### PR TITLE
Make use of MAP_EXCL on FreeBSD.

### PR DESCRIPTION
**Pull request by trasz**
The "O_FIXED | MAP_EXCL" combination works just like MAP_FIXED_NOREPLACE
recently introduced in Linux - except it's for FreeBSD and was added
some four years ago.

### CODE REVIEW

**Comment by trasz** [comment]
Ping?

**Comment by trasz** [comment]
Okay, so this results in something I didn't quite expect; it looks like this (it's the diff of truss(1) (think strace) output between python running with the vanilla jemalloc vs the same binary with patched jemalloc:

```
 mmap(0x0,262144,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34391683072 (0x801e77000)
-mmap(0x801e77000,65536,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34391945216 (0x801eb7000)
-munmap(0x801eb7000,65536)                       = 0 (0x0)
-mmap(0x0,528384,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34391945216 (0x801eb7000)
+mmap(0x801e77000,65536,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON|MAP_EXCL,-1,0x0) ERR#12 'Cannot allocate memory'
+mmap(0x0,528384,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_ALIGNED(12),-1,0x0) = 34391945216 (0x801eb7000)
 mmap(0x0,262144,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34392473600 (0x801f38000)
-mmap(0x801f38000,131072,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34392735744 (0x801f78000)
-munmap(0x801f78000,131072)                      = 0 (0x0)
-mmap(0x0,659456,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34392735744 (0x801f78000)
+mmap(0x801f38000,131072,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|MAP_ANON|MAP_EXCL,-1,0x0) ERR#12 'Cannot allocate memory'
+mmap(0x0,659456,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_ALIGNED(12),-1,0x0) = 34392735744 (0x801f78000)
 mmap(0x0,262144,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0x0) = 34393395200 (0x802019000)

```

... and so on.  So, the patch does improve things by reducing the number of syscalls (and the number of actual map modifications), but why does jemalloc request ranges that are obviously busy, because they were returned by the previous call to mmap?

**Comment by trasz** [comment]
Ping?  The #1260 isn't a blocker for this one; it works just fine even with the bug.

**Comment by davidtgoldblatt** [comment]
Sorry about the delay here -- we're tracking down a couple of slight regressions, and so don't want to add even more range to the search.

**Comment by trasz** [comment]
Ping again?

File: src/pages.c [hunk 1]
```diff
@@ -180,6 +181,29 @@ pages_map(void *addr, size_t size, size_t alignment, bool *commit) {
...
+	 * touching existing mappings, and to mmap with specific alignment.
+	 */
+	{
+		int prot = *commit ? PAGES_PROT_COMMIT : PAGES_PROT_DECOMMIT;
+		int flags = mmap_flags;
+
+		if (addr != NULL) {
+			flags |= MAP_FIXED | MAP_EXCL;
+		} else {
+			flags |= MAP_ALIGNED(ffs(alignment) - 1);
```
Comment (davidtgoldblatt) [inline src/pages.c]: I think this should be `ffs_zu` (and that, therefore, we don't need to include string.h above.

File: src/pages.c [hunk 2]
```diff
@@ -193,7 +193,10 @@ pages_map(void *addr, size_t size, size_t alignment, bool *commit) {
 		if (addr != NULL) {
 			flags |= MAP_FIXED | MAP_EXCL;
 		} else {
-			flags |= MAP_ALIGNED(ffs(alignment) - 1);
+			int alignment_bits = ffs(alignment);
```
Comment (davidtgoldblatt) [inline src/pages.c]: Ditto with `ffs_zu`.

Probably even better, we should assert the alignment; we should always have tweaked it as necessary before we get to the pages module.

**Comment by trasz** [comment]
Done.

**Comment by trasz** [comment]
Ping?  :-)

**Comment by davidtgoldblatt** [comment]
Looks great, thanks!

**Pull request merged**